spec/Section 3 -- Type System.md

```diff
@@ -1580,6 +1583,17 @@ Literal Value            | Variables               | Coerced Value
       characters {"__"} (two underscores).
    3. The input field must accept a type where {IsInputType(inputFieldType)}
       returns {true}.
+   4. If the input field has a non-null default value:
+      1. If the input field references this Input Object either directly or
+         through referenced Input Objects, all input fields in the chain of
+         references which reference this Input Object must either:
+         1. have no default value; or
+         2. have a {null} default value; or
+         3. have a default value, {nestedDefaultValue}, such that the value for
+            this field within {nestedDefaultValue} is either {null} or an empty
+            list.
```

Comment:
Could this possibly be incorporated into the existing step 3 below? 

Perhaps:

> at least one of the fields in the chain of references must be either a nullable or a List type.

adds something along the lines of "and have either no default value, or a default value of either {null} or an empty list"

Reply:
I don't think so, these constraints live at different levels. Step 3 below relates to the entire "default value" either being not present, null, or an empty list; however the text here (above) relates to the _fields_ of the object within the default value. E.g. the following example:

```graphql
input A {
  b: B = {}
}

input B {
  c: C = { a: null }
}

input C {
  a: A = {}
}
```

is fine since no matter whether we reference type A, B or C initially the chain will terminate at C.a since the default for B.c includes that `C.a` is null. Note: at no point is the default value for a field null or an empty list.

The wording could definitely do with some work though :sweat_smile:

Reply:
Scratch that; I think you're right but the wording you've suggested doesn't quite work (see above example). Maybe:

3. If an Input Object references itself either directly or through referenced
   Input Objects:
   1. at least one of the fields in the chain of references must be
     either a nullable or a List type.
   2. at least one of the fields in the chain must have no default value, have a default value of either {null} or an empty list, or have a default value that results in all instances of one of the fields in the chain defaulting to {null} or an empty list)


Bit awkward wording, but hopefully you catch my drift. i.e. allow the following:

```
input A {
  b: B! = {}
}

input B {
  c: C = {}
}

input C {
  a: A! = { b: { c: null} }
}
```

Reply:
These are very helpful examples! I was definitely too restrictive in my rule set. I think what I currently have implemented in PR would reject all cycles, but would also incorrectly reject this example.

I agree that wording isn't quite right, but it's much closer. I like that this has just one rule about circular references but it makes it clear that there are multiple ways to have illegal circular references - and the above rule can specifically be about type checking the default value

Reply:
just to be clear here:
the coerced default value for `A` =  `{b: {c: {a: {b: {c: null}}}}}`.
the coerced default value for `B` =  `{c: {a: {b: {c: null}}}}`.
the coerced default value for `C` =  `{a: {b: {c: null}}}`

That correct?

Reply:
also: we would have to apply this validation rule also for default variable values, right?

Reply:
I don't think we also need it for variable values or argument defaults since those can only be the entry to a circular ref but not actually contained within a circular ref.

```graphql
input A {
  a: A = {}
}

query Test ($x: A = {}) {}
```

This is problematic because `A.a` creates a circular reference, but `$x` only references it, it isn't contained in it. I think more generally you can say that given a set of Input Objects that have no circular reference, there exists no default value for arguments or variable values that would create a circular reference.

Reply:
> just to be clear here:
> the coerced default value for `A` = `{b: {c: {a: {b: {c: null}}}}}`.
> the coerced default value for `B` = `{c: {a: {b: {c: null}}}}`.
> the coerced default value for `C` = `{a: {b: {c: null}}}`
> 
> That correct?

Coerced default values apply to fields, not types - so:

> the coerced default value for `A.b` = `{c: {a: {b: {c: null}}}}`.
> the coerced default value for `B.c` = `{a: {b: {c: null}}}`.
> the coerced default value for `C.a` = `{b: {c: null}}`

Reply:
Okay I've had a go at writing an algorithm for this; I think the key learning whilst attempting this was that the _act of referencing the default value a second time_ is what triggers a cycle; so the algoritm only tracks when the default value of a field is referenced.